processor/src/main/java/com/linecorp/decaton/processor/processors/BatchingProcessor.java

```diff
@@ -0,0 +1,137 @@
+import com.linecorp.decaton.processor.DeferredCompletion;
+import com.linecorp.decaton.processor.ProcessingContext;
+import com.linecorp.decaton.processor.Completion;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.experimental.Accessors;
+
+abstract public class BatchingProcessor<T> implements DecatonProcessor<T> {
+
+    private final ScheduledExecutorService executor;
+    private final List<BatchingTask<T>> windowedTasks = Collections.synchronizedList(new ArrayList<>());
+    private final long lingerMillis;
+    private final int capacity;
+
+    @Getter
+    @Accessors(fluent = true)
+    public static class BatchingTask<T> {
+        @Getter(AccessLevel.NONE)
+        public final Completion completion;
+        public final ProcessingContext<T> context;
+        public final T task;
+
+        private BatchingTask(Completion completion, ProcessingContext<T> context, T task) {
+            this.completion = completion;
+            this.context = context;
+            this.task = task;
+        }
+    }
+
+    // visible for testing
+    BatchingProcessor(
+        long lingerMillis,
+        int capacity,
+        ScheduledThreadPoolExecutor scheduledExecutor
+    ) {
+        this.lingerMillis = lingerMillis;
+        this.capacity = capacity;
+
+        if (scheduledExecutor == null) {
+            scheduledExecutor = new ScheduledThreadPoolExecutor(1, r -> {
+                Thread th = new Thread(r);
+                th.setName("Decaton" + BatchingProcessor.class.getSimpleName()
+                           + '/' + System.identityHashCode(this));
+                return th;
+            });
+        }
+        // For faster shutdown cancel all pending flush on executor shutdown.
+        // In fact for this purpose we have two options here:
+        // A. Cancel all pending flush
+        // B. Run all pending flush immediately, regardless remaining time until flush
+        // A has a downside that it forces consumer to re-process some amount of tasks.
+        // B has a downside that it may causes flush bursting and then on spike for storage/API access in
+        // downstream processor.
+        // Given that this feature is expected to be used mainly for reducing workload for storage/APIs,
+        // we think taking A is much desirable behavior in most cases.
+        scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor = scheduledExecutor;
+    }
+
+    public BatchingProcessor(long lingerMillis) {
+        this(lingerMillis, Integer.MAX_VALUE, null);
+    }
+
+    public BatchingProcessor(long lingerMillis, int capacity) {
+        this(lingerMillis, capacity, null);
+    }
+
+    private void flush() {
+        if (!windowedTasks.isEmpty()) {
+            return;
+        }
+        synchronized (windowedTasks) {
+            processBatchingTasks(windowedTasks);
+        }
+    }
+
+    // visible for testing
+    Runnable flushTask() {
+        return this::flush;
```

Comment:
Seems flushTask() isn't called periodically?

Reply:
I was thinking about the following, but it was certainly difficult to understand, so I fixed it!
1. When the first task comes in, `scheduleFlush()` is called.
2. `flush()` empties windowsTasks.
3. When the next task comes in after `flush()`,  `scheduleFlush()` is called because windowTasks is empty.
2ef1bc0